File: src/spiralverse/temporal_intent.py
"""
Temporal-Intent Harmonic Scaling - H_eff(d,R,x) Extended Harmonic Wall
======================================================================

Extends the Harmonic Scaling Law with temporal intent accumulation:

    H_eff(d, R, x) = R^(d^2 * x)

Where:
    d = distance from safe operation (Poincare ball, Layer 5)
    R = harmonic ratio (1.5 = perfect fifth)
    x = temporal intent factor derived from existing Layer 11 + CPSE channels

The 'x' factor is NOT a new concept - it aggregates existing metrics:

    x(t) = f(d_tri(t), chaosdev(t), fractaldev(t), energydev(t))

Where:
    d_tri(t)      = Triadic temporal distance (L11: immediate/medium/long-term)
    chaosdev(t)   = Lyapunov-based chaos deviation (CPSE z-vector)
    fractaldev(t) = Fractal dimension deviation
    energydev(t)  = Energy channel deviation

This makes security cost compound based on SUSTAINED adversarial behavior,
not just instantaneous distance. Brief deviations are forgiven; persistent
drift toward the boundary costs super-exponentially more over time.

Integration with existing layers:
    - L5:  Hyperbolic distance provides 'd'
    - L11: Triadic Temporal Distance provides d_tri(t)
    - L12: Harmonic Wall now uses H_eff(d,R,x) instead of H(d,R)
    - CPSE: Chaos/fractal/energy deviation channels provide z_t

From requirements.md AC-2.3.3:
    Omega = pqc_valid * harm_score * (1 - drift_norm/drift_max) * triadic_stable * spectral_score

The 'x' factor integrates drift_norm accumulation over time windows, keeping
everything axiom-safe with Layer 11 (Triadic Temporal) and CPSE z-vector tests.

"Security IS growth. Intent over time reveals truth."
"""

import math
import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple
from enum import Enum
from collections import deque
import hashlib


# =============================================================================
# Constants
# =============================================================================

# Harmonic ratio (perfect fifth)
R_HARMONIC = 1.5

# Intent decay rate (how fast old intent fades)
INTENT_DECAY_RATE = 0.95  # per time window

# Time window for intent accumulation (seconds)
INTENT_WINDOW_SECONDS = 1.0

# Maximum intent accumulation before hard exile
MAX_INTENT_ACCUMULATION = 10.0

# Trust threshold for exile (from AC-2.3.2)
TRUST_EXILE_THRESHOLD = 0.3
TRUST_EXILE_ROUNDS = 10


# =============================================================================
# Intent State
# =============================================================================

class IntentState(Enum):
    """Classification of agent's temporal intent."""
    BENIGN = "benign"           # x < 0.5 - consistently safe
    NEUTRAL = "neutral"         # 0.5 <= x < 1.0 - normal operation
    DRIFTING = "drifting"       # 1.0 <= x < 2.0 - concerning pattern
    ADVERSARIAL = "adversarial" # x >= 2.0 - sustained adversarial behavior
    EXILED = "exiled"           # Null-space exile triggered


@dataclass
class IntentSample:
    """
    Single sample of distance/intent at a point in time.

    Integrates L5 (hyperbolic distance), L11 (triadic temporal), and CPSE z-vector.
    """
    timestamp: float
    distance: float        # d in Poincare ball (0 to ~1) from L5
    velocity: float        # rate of change of distance
    harmony: float         # CHARM value (-1 to 1)

    # CPSE z-vector deviation channels
    chaosdev: float = 0.0    # Lyapunov-based chaos deviation
    fractaldev: float = 0.0  # Fractal dimension deviation
    energydev: float = 0.0   # Energy channel deviation

    # Triadic temporal components (L11)
    d_tri_immediate: float = 0.0   # Immediate behavior
    d_tri_medium: float = 0.0      # Medium-term pattern
    d_tri_long: float = 0.0        # Long-term trajectory

    @property
    def d_tri(self) -> float:
        """Triadic temporal distance (L11) - geometric mean of time scales."""
        return (abs(self.d_tri_immediate) * abs(self.d_tri_medium) * abs(self.d_tri_long)) ** (1/3)

    @property
    def raw_intent(self) -> float:
        """
        Calculate raw intent from this sample using existing L11 + CPSE metrics.

        x(t) = f(d_tri(t), chaosdev(t), fractaldev(t), energydev(t))
        """
        # Velocity contribution (moving toward boundary is adversarial)
        velocity_factor = max(0, self.velocity) * 2.0

        # Distance contribution (further out = more suspicious)
        distance_factor = self.distance ** 2

        # Harmony dampening (high harmony reduces intent score)
        harmony_dampening = (1 - self.harmony) / 2  # 0 to 1

        # CPSE deviation channels contribution
        cpse_factor = (abs(self.chaosdev) + abs(self.fractaldev) + abs(self.energydev)) / 3

        # Triadic temporal contribution (L11)
        triadic_factor = self.d_tri

        base_intent = (velocity_factor + distance_factor) * (0.5 + harmony_dampening)

        # Amplify by CPSE deviations and triadic distance
        return base_intent * (1.0 + cpse_factor + triadic_factor)


@dataclass
class IntentHistory:
    """
    Tracks intent accumulation over time for a single agent.

    Uses sliding window to accumulate intent, with decay for old samples.
    """
    agent_id: str
    samples: deque = field(default_factory=lambda: deque(maxlen=1000))
    accumulated_intent: float = 0.0
    trust_score: float = 1.0
    low_trust_rounds: int = 0
    state: IntentState = IntentState.NEUTRAL
    last_update: float = field(default_factory=time.time)

    def add_sample(self, distance: float, velocity: float = 0.0, harmony: float = 0.0):
        """Add a new intent sample and update accumulation."""
        now = time.time()

        sample = IntentSample(
            timestamp=now,
            distance=distance,
            velocity=velocity,
            harmony=harmony
        )
        self.samples.append(sample)

        # Apply decay to accumulated intent
        time_delta = now - self.last_update
        decay_factor = INTENT_DECAY_RATE ** (time_delta / INTENT_WINDOW_SECONDS)
        self.accumulated_intent *= decay_factor

        # Add new intent
        self.accumulated_intent += sample.raw_intent

        # Cap at maximum
        self.accumulated_intent = min(self.accumulated_intent, MAX_INTENT_ACCUMULATION)

        # Update trust score
        self._update_trust()

        # Update state classification
        self._update_state()

        self.last_update = now

    def _update_trust(self):
        """Update trust score based on recent behavior."""
        if len(self.samples) < 5:
            return

        # Average recent distances
        recent = list(self.samples)[-10:]
        avg_distance = sum(s.distance for s in recent) / len(recent)

        # Trust decreases with distance and accumulated intent
        trust_change = -0.1 * avg_distance - 0.05 * self.accumulated_intent

        # Trust recovers slowly when behaving well
        if self.accumulated_intent < 0.5 and avg_distance < 0.3:
            trust_change += 0.02

        self.trust_score = max(0.0, min(1.0, self.trust_score + trust_change))

        # Track low trust rounds for exile
        if self.trust_score < TRUST_EXILE_THRESHOLD:
            self.low_trust_rounds += 1
        else:
            self.low_trust_rounds = 0

    def _update_state(self):
        """Classify intent state based on accumulation."""
        x = self.accumulated_intent

        # Check for exile condition (AC-2.3.2)
        if self.low_trust_rounds >= TRUST_EXILE_ROUNDS:
            self.state = IntentState.EXILED
        elif x < 0.5:
            self.state = IntentState.BENIGN
        elif x < 1.0:
            self.state = IntentState.NEUTRAL
        elif x < 2.0:
            self.state = IntentState.DRIFTING
        else:
            self.state = IntentState.ADVERSARIAL

    @property
    def x_factor(self) -> float:
        """
        Get the 'x' factor for H(d,R)^x formula.

        Returns a value typically between 0.5 and 3.0:
        - x < 1: Forgiving (brief deviation)
        - x = 1: Standard H(d,R)
        - x > 1: Compounding (sustained adversarial)
        """
        # Base x from accumulated intent
        base_x = 0.5 + self.accumulated_intent * 0.25

        # Modify by trust score (low trust amplifies x)
        trust_modifier = 1.0 + (1.0 - self.trust_score)

        return min(3.0, base_x * trust_modifier)


# =============================================================================
# Extended Harmonic Wall
# =============================================================================

def harmonic_wall_basic(d: float, R: float = R_HARMONIC) -> float:
    """
    Original Harmonic Wall: H(d, R) = R^(d²)
    """
    return R ** (d ** 2)


def harmonic_wall_temporal(d: float, x: float, R: float = R_HARMONIC) -> float:
    """
    Extended Harmonic Wall with temporal intent: H(d, R)^x = R^(d² · x)

    Args:
        d: Distance from safe operation (0 to ~1 in Poincaré ball)
        x: Intent persistence factor from IntentHistory.x_factor
        R: Harmonic ratio (default 1.5 = perfect fifth)

    Returns:
        Security cost multiplier (grows superexponentially with sustained drift)
    """
    return R ** (d ** 2 * x)


def compare_scaling(d: float, x: float) -> Dict[str, float]:
    """
    Compare basic vs temporal harmonic wall at given distance and intent.
    """
    basic = harmonic_wall_basic(d)
    temporal = harmonic_wall_temporal(d, x)

    return {
        "distance": d,
        "x_factor": x,
        "H_basic": basic,
        "H_temporal": temporal,
        "amplification": temporal / basic if basic > 0 else float('inf')
    }


# =============================================================================
# Security Gate Integration
# =============================================================================

@dataclass
class TemporalSecurityGate:
    """
    Security gate that uses H(d,R)^x for authorization decisions.

    Integrates with the Omega decision function (AC-2.3.3):
    Ω = pqc_valid × harm_score × (1 - drift_norm/drift_max) × triadic_stable × spectral_score

    The temporal intent scaling modifies harm_score via H(d,R)^x.
    """
    histories: Dict[str, IntentHistory] = field(default_factory=dict)

    # Decision thresholds from AC-2.3.4
    ALLOW_THRESHOLD = 0.85
    QUARANTINE_THRESHOLD = 0.40

    def get_or_create_history(self, agent_id: str) -> IntentHistory:
        """Get or create intent history for an agent."""
        if agent_id not in self.histories:
            self.histories[agent_id] = IntentHistory(agent_id=agent_id)
        return self.histories[agent_id]

    def record_observation(
        self,
        agent_id: str,
        distance: float,
        velocity: float = 0.0,
        harmony: float = 0.0
    ):
        """Record an observation for an agent."""
        history = self.get_or_create_history(agent_id)
        history.add_sample(distance, velocity, harmony)

    def compute_omega(
        self,
        agent_id: str,
        pqc_valid: bool = True,
        triadic_stable: float = 1.0,
        spectral_score: float = 1.0
    ) -> Tuple[float, str]:
        """
        Compute Omega decision score using temporal intent scaling.

        Ω = pqc_valid × harm_score × drift_factor × triadic_stable × spectral_score

        Where harm_score = 1 / H(d, R)^x (inverted so higher is better)
        And drift_factor = (1 - drift_norm/drift_max)

        Returns:
            (omega_score, decision) where decision is ALLOW/QUARANTINE/DENY/EXILE
        """
        history = self.get_or_create_history(agent_id)

        # Check for exile
        if history.state == IntentState.EXILED:
            return 0.0, "EXILE"

        # Get latest distance
        if history.samples:
            latest = history.samples[-1]
            d = latest.distance
        else:
            d = 0.0

        # Compute H(d,R)^x
        x = history.x_factor
        h_temporal = harmonic_wall_temporal(d, x)

        # Invert for harm_score (lower H = higher score = safer)
        # Use 1/(1 + log(H)) to keep in reasonable range
        harm_score = 1.0 / (1.0 + math.log(max(1.0, h_temporal)))

        # Drift factor from accumulated intent
        drift_factor = 1.0 - (history.accumulated_intent / MAX_INTENT_ACCUMULATION)

        # PQC factor
        pqc_factor = 1.0 if pqc_valid else 0.0

        # Compute Omega
        omega = pqc_factor * harm_score * drift_factor * triadic_stable * spectral_score

        # Decision
        if omega > self.ALLOW_THRESHOLD:
            decision = "ALLOW"
        elif omega > self.QUARANTINE_THRESHOLD:
            decision = "QUARANTINE"
        else:
            decision = "DENY"

        return omega, decision

    def get_status(self, agent_id: str) -> Dict:
        """Get full status for an agent."""
        history = self.get_or_create_history(agent_id)

        omega, decision = self.compute_omega(agent_id)

        return {
            "agent_id": agent_id,
            "state": history.state.value,
            "trust_score": history.trust_score,
            "accumulated_intent": history.accumulated_intent,
            "x_factor": history.x_factor,
            "low_trust_rounds": history.low_trust_rounds,
            "samples": len(history.samples),
            "omega": omega,
            "decision": decision
        }


# =============================================================================
# Demo
# =============================================================================

def demo():
    """Demonstrate temporal intent scaling."""
    print("=" * 70)
    print("  TEMPORAL INTENT SCALING - H(d,R)^x Extended Harmonic Wall")
    print("  'Security IS growth. Intent over time reveals truth.'")
    print("=" * 70)
    print()

    # Demo 1: Compare basic vs temporal at various distances and intents
    print("[SCALING] H(d,R) vs H(d,R)^x comparison:")
    print("-" * 60)
    print(f"  {'Distance':>8} {'x_factor':>8} {'H_basic':>12} {'H_temporal':>12} {'Amplify':>10}")
    print("-" * 60)

    test_cases = [
        (0.3, 0.5),   # Low distance, brief deviation (forgiving)
        (0.3, 1.0),   # Low distance, normal
        (0.3, 2.0),   # Low distance, sustained intent
        (0.6, 0.5),   # Medium distance, brief
        (0.6, 1.0),   # Medium distance, normal
        (0.6, 2.0),   # Medium distance, sustained
        (0.9, 0.5),   # High distance, brief
        (0.9, 1.0),   # High distance, normal
        (0.9, 2.0),   # High distance, sustained (catastrophic)
    ]

    for d, x in test_cases:
        result = compare_scaling(d, x)
        print(f"  {d:>8.2f} {x:>8.2f} {result['H_basic']:>12.2f} {result['H_temporal']:>12.2f} {result['amplification']:>10.2f}x")
    print()

    # Demo 2: Simulate agent behavior over time
    print("[SIMULATION] Agent behavior over time:")
    print("-" * 60)

    gate = TemporalSecurityGate()

    # Simulate benign agent
    print("  Benign Agent (stays near center):")
    for i in range(10):
        gate.record_observation("benign", distance=0.1 + 0.05 * (i % 3), harmony=0.8)
    status = gate.get_status("benign")
    print(f"    State: {status['state']}, x={status['x_factor']:.2f}, Omega={status['omega']:.3f} -> {status['decision']}")

    # Simulate drifting agent
    print("  Drifting Agent (gradually moves toward boundary):")
    for i in range(15):
        gate.record_observation("drifter", distance=0.2 + 0.04 * i, velocity=0.04, harmony=0.3)
    status = gate.get_status("drifter")
    print(f"    State: {status['state']}, x={status['x_factor']:.2f}, Omega={status['omega']:.3f} -> {status['decision']}")

    # Simulate adversarial agent
    print("  Adversarial Agent (sustained boundary approach):")
    for i in range(20):
        gate.record_observation("adversary", distance=0.7 + 0.01 * i, velocity=0.1, harmony=-0.5)
    status = gate.get_status("adversary")
    print(f"    State: {status['state']}, x={status['x_factor']:.2f}, Omega={status['omega']:.3f} -> {status['decision']}")

    # Simulate recovered agent (was drifting, came back)
    print("  Recovered Agent (drifted then returned):")
    for i in range(10):
        gate.record_observation("recovered", distance=0.5 + 0.03 * i, velocity=0.03, harmony=0.2)
    for i in range(15):
        gate.record_observation("recovered", distance=0.8 - 0.04 * i, velocity=-0.04, harmony=0.7)
    status = gate.get_status("recovered")
    print(f"    State: {status['state']}, x={status['x_factor']:.2f}, Omega={status['omega']:.3f} -> {status['decision']}")

    print()

    # Demo 3: Security amplification at boundary
    print("[BOUNDARY] Security amplification near Poincaré boundary:")
    print("-" * 60)
    print("  At d=0.95 (very close to boundary):")
    for x in [0.5, 1.0, 1.5, 2.0, 2.5, 3.0]:
        h = harmonic_wall_temporal(0.95, x)
        print(f"    x={x:.1f}: H(0.95, 1.5)^{x:.1f} = {h:,.0f}x security cost")
    print()

    print("=" * 70)
    print("  Summary:")
    print("    - Brief deviations (x<1): Forgiving, allows recovery")
    print("    - Normal operation (x=1): Standard H(d,R) scaling")
    print("    - Sustained drift (x>1): Compounding costs, triggers quarantine")
    print("    - Persistent adversarial (x>2): Near-infinite cost, exile imminent")
    print()
    print("  Formula: H(d, R)^x = R^(d² · x) where R=1.5")
    print("=" * 70)


if __name__ == "__main__":
    demo()

File: src/harmonic/temporalIntent.ts
/**
 * @file temporalIntent.ts
 * @module harmonic/temporalIntent
 * @layer Layer 5, Layer 11, Layer 12, Layer 13
 * @component Temporal-Intent Harmonic Scaling
 * @version 3.2.4
 *
 * Extends the Harmonic Scaling Law with temporal intent accumulation:
 *
 *     H_eff(d, R, x) = R^(d² · x)
 *
 * Where:
 *     d = distance from safe operation (Poincaré ball, Layer 5)
 *     R = harmonic ratio (1.5 = perfect fifth)
 *     x = temporal intent factor derived from L11 + CPSE channels
 *
 * The 'x' factor aggregates existing metrics:
 *
 *     x(t) = f(d_tri(t), chaosdev(t), fractaldev(t), energydev(t))
 *
 * This makes security cost compound based on SUSTAINED adversarial behavior,
 * not just instantaneous distance. Brief deviations are forgiven; persistent
 * drift toward the boundary costs super-exponentially more over time.
 *
 * Integration with existing layers:
 *     - L5:  Hyperbolic distance provides 'd'
 *     - L11: Triadic Temporal Distance provides d_tri(t)
 *     - L12: Harmonic Wall now uses H_eff(d,R,x) instead of H(d,R)
 *     - CPSE: Chaos/fractal/energy deviation channels provide z_t
 *
 * "Security IS growth. Intent over time reveals truth."
 */

import type { Vector6D } from './constants.js';

// ═══════════════════════════════════════════════════════════════
// Constants
// ═══════════════════════════════════════════════════════════════

/** Harmonic ratio (perfect fifth) */
export const R_HARMONIC = 1.5;

/** Intent decay rate per time window (how fast old intent fades) */
export const INTENT_DECAY_RATE = 0.95;

/** Time window for intent accumulation (seconds) */
export const INTENT_WINDOW_SECONDS = 1.0;

/** Maximum intent accumulation before hard exile */
export const MAX_INTENT_ACCUMULATION = 10.0;

/** Trust threshold for exile (from AC-2.3.2) */
export const TRUST_EXILE_THRESHOLD = 0.3;

/** Consecutive low-trust rounds to trigger exile */
export const TRUST_EXILE_ROUNDS = 10;

// ═══════════════════════════════════════════════════════════════
// Intent State
// ═══════════════════════════════════════════════════════════════

/** Classification of agent's temporal intent */
export enum IntentState {
  /** x < 0.5 — consistently safe */
  BENIGN = 'benign',
  /** 0.5 <= x < 1.0 — normal operation */
  NEUTRAL = 'neutral',
  /** 1.0 <= x < 2.0 — concerning pattern */
  DRIFTING = 'drifting',
  /** x >= 2.0 — sustained adversarial behavior */
  ADVERSARIAL = 'adversarial',
  /** Null-space exile triggered */
  EXILED = 'exiled',
}

// ═══════════════════════════════════════════════════════════════
// Intent Sample
// ═══════════════════════════════════════════════════════════════

/** Single sample of distance/intent at a point in time */
export interface IntentSampleInput {
  /** Timestamp in ms (Date.now() compatible) */
  timestamp: number;
  /** Distance in Poincaré ball (0 to ~1) from L5 */
  distance: number;
  /** Rate of change of distance */
  velocity: number;
  /** CHARM value (-1 to 1) */
  harmony: number;
  /** Lyapunov-based chaos deviation (CPSE z-vector) */
  chaosdev?: number;
  /** Fractal dimension deviation */
  fractaldev?: number;
  /** Energy channel deviation */
  energydev?: number;
  /** Triadic temporal: immediate behavior */
  dTriImmediate?: number;
  /** Triadic temporal: medium-term pattern */
  dTriMedium?: number;
  /** Triadic temporal: long-term trajectory */
  dTriLong?: number;
}

/** Computed intent sample with derived metrics */
export interface IntentSample extends IntentSampleInput {
  /** Triadic temporal distance (L11) — geometric mean of 3 scales */
  dTri: number;
  /** Raw intent value for this sample */
  rawIntent: number;
}

/**
 * Compute triadic temporal distance (L11) — geometric mean of 3 time scales.
 */
export function computeDTri(
  immediate: number,
  medium: number,
  long: number
): number {
  return Math.cbrt(Math.abs(immediate) * Math.abs(medium) * Math.abs(long));
}

/**
 * Compute raw intent from a single sample using L11 + CPSE metrics.
 *
 * x(t) = f(d_tri(t), chaosdev(t), fractaldev(t), energydev(t))
 */
export function computeRawIntent(input: IntentSampleInput): number {
  const chaosdev = input.chaosdev ?? 0;
  const fractaldev = input.fractaldev ?? 0;
  const energydev = input.energydev ?? 0;
  const dTriImmediate = input.dTriImmediate ?? 0;
  const dTriMedium = input.dTriMedium ?? 0;
  const dTriLong = input.dTriLong ?? 0;

  // Velocity contribution (moving toward boundary is adversarial)
  const velocityFactor = Math.max(0, input.velocity) * 2.0;

  // Distance contribution (further out = more suspicious)
  const distanceFactor = input.distance ** 2;

  // Harmony dampening (high harmony reduces intent score)
  const harmonyDampening = (1 - input.harmony) / 2; // 0 to 1

  // CPSE deviation channels contribution
  const cpseFactor = (Math.abs(chaosdev) + Math.abs(fractaldev) + Math.abs(energydev)) / 3;

  // Triadic temporal contribution (L11)
  const triadicFactor = computeDTri(dTriImmediate, dTriMedium, dTriLong);

  const baseIntent = (velocityFactor + distanceFactor) * (0.5 + harmonyDampening);

  // Amplify by CPSE deviations and triadic distance
  return baseIntent * (1.0 + cpseFactor + triadicFactor);
}

/**
 * Build a full IntentSample from input, computing derived metrics.
 */
export function buildSample(input: IntentSampleInput): IntentSample {
  return {
    ...input,
    chaosdev: input.chaosdev ?? 0,
    fractaldev: input.fractaldev ?? 0,
    energydev: input.energydev ?? 0,
    dTriImmediate: input.dTriImmediate ?? 0,
    dTriMedium: input.dTriMedium ?? 0,
    dTriLong: input.dTriLong ?? 0,
    dTri: computeDTri(
      input.dTriImmediate ?? 0,
      input.dTriMedium ?? 0,
      input.dTriLong ?? 0
    ),
    rawIntent: computeRawIntent(input),
  };
}

// ═══════════════════════════════════════════════════════════════
// Intent History
// ═══════════════════════════════════════════════════════════════

/** Full state of an agent's intent history */
export interface IntentHistoryState {
  agentId: string;
  samples: IntentSample[];
  accumulatedIntent: number;
  trustScore: number;
  lowTrustRounds: number;
  state: IntentState;
  lastUpdateMs: number;
}

/** Maximum samples to retain */
const MAX_SAMPLES = 1000;

/**
 * Create a fresh intent history for an agent.
 */
export function createIntentHistory(agentId: string, nowMs?: number): IntentHistoryState {
  return {
    agentId,
    samples: [],
    accumulatedIntent: 0,
    trustScore: 1.0,
    lowTrustRounds: 0,
    state: IntentState.NEUTRAL,
    lastUpdateMs: nowMs ?? Date.now(),
  };
}

/**
 * Add a sample to intent history, updating accumulation, trust, and state.
 * Returns a new IntentHistoryState (immutable update).
 */
export function addSample(
  history: IntentHistoryState,
  distance: number,
  velocity: number = 0,
  harmony: number = 0,
  nowMs?: number
): IntentHistoryState {
  const now = nowMs ?? Date.now();

  const sample = buildSample({
    timestamp: now,
    distance,
    velocity,
    harmony,
  });

  // Append sample (cap at MAX_SAMPLES)
  const samples =
    history.samples.length >= MAX_SAMPLES
      ? [...history.samples.slice(1), sample]
      : [...history.samples, sample];

  // Apply decay to accumulated intent
  const timeDeltaSec = (now - history.lastUpdateMs) / 1000;
  const decayFactor = Math.pow(INTENT_DECAY_RATE, timeDeltaSec / INTENT_WINDOW_SECONDS);
  let accumulatedIntent = history.accumulatedIntent * decayFactor + sample.rawIntent;
  accumulatedIntent = Math.min(accumulatedIntent, MAX_INTENT_ACCUMULATION);

  // Update trust score
  let trustScore = history.trustScore;
  let lowTrustRounds = history.lowTrustRounds;
  if (samples.length >= 5) {
    const recent = samples.slice(-10);
    const avgDistance = recent.reduce((s, s2) => s + s2.distance, 0) / recent.length;
    let trustChange = -0.1 * avgDistance - 0.05 * accumulatedIntent;
    if (accumulatedIntent < 0.5 && avgDistance < 0.3) {
      trustChange += 0.02;
    }
    trustScore = Math.max(0, Math.min(1, trustScore + trustChange));
  }

  if (trustScore < TRUST_EXILE_THRESHOLD) {
    lowTrustRounds = lowTrustRounds + 1;
  } else {
    lowTrustRounds = 0;
  }

  // Classify intent state
  let state: IntentState;
  if (lowTrustRounds >= TRUST_EXILE_ROUNDS) {
    state = IntentState.EXILED;
  } else if (accumulatedIntent < 0.5) {
    state = IntentState.BENIGN;
  } else if (accumulatedIntent < 1.0) {
    state = IntentState.NEUTRAL;
  } else if (accumulatedIntent < 2.0) {
    state = IntentState.DRIFTING;
  } else {
    state = IntentState.ADVERSARIAL;
  }

  return {
    agentId: history.agentId,
    samples,
    accumulatedIntent,
    trustScore,
    lowTrustRounds,
    state,
    lastUpdateMs: now,
  };
}

/**
 * Compute the x-factor for H(d,R)^x from an intent history.
 *
 * Returns a value typically between 0.5 and 3.0:
 *   - x < 1: Forgiving (brief deviation)
 *   - x = 1: Standard H(d,R)
 *   - x > 1: Compounding (sustained adversarial)
 */
export function computeXFactor(history: IntentHistoryState): number {
  const baseX = 0.5 + history.accumulatedIntent * 0.25;
  const trustModifier = 1.0 + (1.0 - history.trustScore);
  return Math.min(3.0, baseX * trustModifier);
}

// ═══════════════════════════════════════════════════════════════
// Extended Harmonic Wall
// ═══════════════════════════════════════════════════════════════

/**
 * Original Harmonic Wall: H(d, R) = R^(d²)
 */
export function harmonicWallBasic(d: number, R: number = R_HARMONIC): number {
  return Math.pow(R, d * d);
}

/**
 * Extended Harmonic Wall with temporal intent: H_eff(d, R, x) = R^(d² · x)
 *
 * @param d — Distance from safe operation (0 to ~1 in Poincaré ball)
 * @param x — Intent persistence factor from computeXFactor
 * @param R — Harmonic ratio (default 1.5 = perfect fifth)
 * @returns Security cost multiplier (grows super-exponentially with sustained drift)
 */
export function harmonicWallTemporal(d: number, x: number, R: number = R_HARMONIC): number {
  return Math.pow(R, d * d * x);
}

/**
 * Compare basic vs temporal harmonic wall at given distance and intent.
 */
export function compareScaling(
  d: number,
  x: number
): { distance: number; xFactor: number; hBasic: number; hTemporal: number; amplification: number } {
  const hBasic = harmonicWallBasic(d);
  const hTemporal = harmonicWallTemporal(d, x);
  return {
    distance: d,
    xFactor: x,
    hBasic,
    hTemporal,
    amplification: hBasic > 0 ? hTemporal / hBasic : Infinity,
  };
}

// ═══════════════════════════════════════════════════════════════
// Temporal Security Gate (Layer 13 Integration)
// ═══════════════════════════════════════════════════════════════

/** Decision thresholds from AC-2.3.4 */
export const ALLOW_THRESHOLD = 0.85;
export const QUARANTINE_THRESHOLD = 0.40;

/** Possible gate decisions */
export type GateDecision = 'ALLOW' | 'QUARANTINE' | 'DENY' | 'EXILE';

/** Result of Omega computation */
export interface OmegaResult {
  omega: number;
  decision: GateDecision;
  xFactor: number;
  hTemporal: number;
  harmScore: number;
  driftFactor: number;
  state: IntentState;
}

/**
 * Compute Omega decision score using temporal intent scaling.
 *
 * Ω = pqc_valid × harm_score × drift_factor × triadic_stable × spectral_score
 *
 * Where:
 *   harm_score  = 1 / (1 + log(H(d, R)^x))   (inverted: higher = safer)
 *   drift_factor = 1 - accumulated_intent / MAX
 */
export function computeOmega(
  history: IntentHistoryState,
  pqcValid: boolean = true,
  triadicStable: number = 1.0,
  spectralScore: number = 1.0
): OmegaResult {
  // Exile check
  if (history.state === IntentState.EXILED) {
    return {
      omega: 0,
      decision: 'EXILE',
      xFactor: computeXFactor(history),
      hTemporal: Infinity,
      harmScore: 0,
      driftFactor: 0,
      state: IntentState.EXILED,
    };
  }

  // Get latest distance
  const d =
    history.samples.length > 0
      ? history.samples[history.samples.length - 1].distance
      : 0;

  const x = computeXFactor(history);
  const hTemporal = harmonicWallTemporal(d, x);

  // Invert for harm_score (lower H = higher score = safer)
  const harmScore = 1.0 / (1.0 + Math.log(Math.max(1.0, hTemporal)));

  // Drift factor from accumulated intent
  const driftFactor = 1.0 - history.accumulatedIntent / MAX_INTENT_ACCUMULATION;

  // PQC factor
  const pqcFactor = pqcValid ? 1.0 : 0.0;

  // Compute Omega
  const omega = pqcFactor * harmScore * driftFactor * triadicStable * spectralScore;

  // Decision
  let decision: GateDecision;
  if (omega > ALLOW_THRESHOLD) {
    decision = 'ALLOW';
  } else if (omega > QUARANTINE_THRESHOLD) {
    decision = 'QUARANTINE';
  } else {
    decision = 'DENY';
  }

  return {
    omega,
    decision,
    xFactor: x,
    hTemporal,
    harmScore,
    driftFactor,
    state: history.state,
  };
}

/**
 * Get full status for an agent's intent history.
 */
export function getStatus(history: IntentHistoryState): {
  agentId: string;
  state: string;
  trustScore: number;
  accumulatedIntent: number;
  xFactor: number;
  lowTrustRounds: number;
  sampleCount: number;
  omega: number;
  decision: GateDecision;
} {
  const result = computeOmega(history);
  return {
    agentId: history.agentId,
    state: history.state,
    trustScore: history.trustScore,
    accumulatedIntent: history.accumulatedIntent,
    xFactor: result.xFactor,
    lowTrustRounds: history.lowTrustRounds,
    sampleCount: history.samples.length,
    omega: result.omega,
    decision: result.decision,
  };
}

// ═══════════════════════════════════════════════════════════════
// Multi-Agent Gate (manages histories for many agents)
// ═══════════════════════════════════════════════════════════════

/** Manages temporal intent histories for multiple agents */
export class TemporalSecurityGate {
  private histories: Map<string, IntentHistoryState> = new Map();

  /** Get or create intent history for an agent */
  getOrCreate(agentId: string, nowMs?: number): IntentHistoryState {
    let h = this.histories.get(agentId);
    if (!h) {
      h = createIntentHistory(agentId, nowMs);
      this.histories.set(agentId, h);
    }
    return h;
  }

  /** Record an observation for an agent */
  recordObservation(
    agentId: string,
    distance: number,
    velocity: number = 0,
    harmony: number = 0,
    nowMs?: number
  ): IntentHistoryState {
    const current = this.getOrCreate(agentId, nowMs);
    const updated = addSample(current, distance, velocity, harmony, nowMs);
    this.histories.set(agentId, updated);
    return updated;
  }

  /** Compute Omega decision for an agent */
  computeOmega(
    agentId: string,
    pqcValid?: boolean,
    triadicStable?: number,
    spectralScore?: number
  ): OmegaResult {
    return computeOmega(this.getOrCreate(agentId), pqcValid, triadicStable, spectralScore);
  }

  /** Get full status for an agent */
  getStatus(agentId: string): ReturnType<typeof getStatus> {
    return getStatus(this.getOrCreate(agentId));
  }

  /** Get all tracked agent IDs */
  agentIds(): string[] {
    return Array.from(this.histories.keys());
  }

  /** Remove an agent's history */
  remove(agentId: string): boolean {
    return this.histories.delete(agentId);
  }
}
